### PR TITLE
Validate http11 body type

### PR DESCRIPTION
**Pull request by AvivC**
Raising `ValueError` in `HTTP11Connection._send_body` on invalid body type

Added tests for the case when the error may be raised. Also testing the
other cases when errors regarding the type of `body` argument may be
raised.
Added to the docstring the option to send an iterable of bytestrings as a body.

### CODE REVIEW

File: test/test_http11.py [hunk 1]
```diff
@@ -446,6 +448,38 @@ def test_http_upgrade_headers_only_sent_once(self):
...
+            c.request('GET', '/get', body=body, headers={'content-length': str(len(str(body)))})
+        assert str('Request body must be a bytestring, a file-like object returning bytestrings '
+                   'or an iterable of bytestrings. Got: {}'.format(type(body))) in str(exc_info)
+
+        rouge_element = 123
+        with pytest.raises(TypeError) as exc_info:
+            body = ['legal1', 'legal2', rouge_element]
+            c.request('GET', '/get', body=body, headers={'content-length': str(len(map(str, body)))})
+        assert str('Elements in iterable body must be bytestrings. Illegal element: {}'
+                   .format(rouge_element)) in str(exc_info)
```
Comment (Lukasa) [inline test/test_http11.py]: Can these be split into two separate tests, please?

File: test/test_http11.py [hunk 2]
```diff
@@ -446,6 +448,38 @@ def test_http_upgrade_headers_only_sent_once(self):
...
+
+    def test_exception_raised_for_filelike_body_not_returning_bytes(self):
+        c = HTTP11Connection('httpbin.org')
+
+        class RougeFile(object):
```
Comment (Lukasa) [inline test/test_http11.py]: Typo: you want `Rogue` here. =)

File: test/test_http11.py [hunk 3]
```diff
@@ -446,6 +448,38 @@ def test_http_upgrade_headers_only_sent_once(self):
...
+
+        # in both cases content-length set so body type is BODY_FLAT. value doesn't matter
+
+        with pytest.raises(TypeError) as exc_info:
+            body = 1234
+            c.request('GET', '/get', body=body, headers={'content-length': str(len(str(body)))})
+        assert str('Request body must be a bytestring, a file-like object returning bytestrings '
+                   'or an iterable of bytestrings. Got: {}'.format(type(body))) in str(exc_info)
+
+        rouge_element = 123
```
Comment (Lukasa) [inline test/test_http11.py]: Typo: you want `rogue_element` here. =)

**Comment by Lukasa** [comment]
Thanks for this!

I've left a few small notes inline that I'd like you to address before I merge this. =) Additionally, I think switching to use a `TypeError` here, while semantically more accurate, is more churn than I want at this stage in the library's life. Do you mind leaving those as `ValueError`? I'll probably end up wanting to introduce a custom exception for this case anyway.

**Comment by AvivC** [comment]
Done :)

**Comment by Lukasa** [comment]
Fantastic @AvivC! Thanks! :sparkles: :cake: :sparkles:

**Comment by AvivC** [comment]
You're welcome :)

**Pull request merged**